In [1]:
import pandas as pd
import numpy as np

In [3]:
df_universal_songs = pd.read_csv("/Users/pratikhotchandani/Downloads/Github/DS_5500/data_collection/cleaned_data/8.csv")

In [4]:
df_universal_songs.head()

Unnamed: 0                track_id  \
0           0  2r0ROhr7pRN4MXDMT1fEmd   
1           1  4I38e6Dg52a2o2a8i5Q5PW   
2           2  59NoiRhnom3lTeRFaBzOev   
3           3  5uUqRQd385pvLxC8JX3tXn   
4           4  1KaBRg2xgNeCljmyxBH1mo   

                                   track_name  \
0                  Leo Das Entry (From "Leo")   
1                                AAO KILLELLE   
2      Mayakiriye Sirikiriye - Orchestral EDM   
3    Scene Ah Scene Ah - Experimental EDM Mix   
4  Gundellonaa X I Am A Disco Dancer - Mashup   

                                         artist_name  year  popularity  \
0                                Anirudh Ravichander  2024          59   
1  Anirudh Ravichander, Pravin Mani, Vaishali Sri...  2024          47   
2           Anirudh Ravichander, Anivee, Alvin Bruno  2024          35   
3  Anirudh Ravichander, Bharath Sankar, Kabilan, ...  2024          24   
4  Anirudh Ravichander, Benny Dayal, Leon James, ...  2024          22   

                                         artwork_url  \
0  https://i.scdn.co/image/ab67616d0000b273ce9c65...   
1  https://i.scdn.co/image/ab67616d0000b273be1b03...   
2  https://i.scdn.co/image/ab67616d0000b27334a1dd...   
3  https://i.scdn.co/image/ab67616d0000b27332e623...   
4  https://i.scdn.co/image/ab67616d0000b2735a59b6...   

                                   album_name  acousticness  danceability  \
0                  Leo Das Entry (From "Leo")        0.0241         0.753   
1                                AAO KILLELLE        0.0851         0.780   
2      Mayakiriye Sirikiriye (Orchestral EDM)        0.0311         0.457   
3    Scene Ah Scene Ah (Experimental EDM Mix)        0.2270         0.718   
4  Gundellonaa X I Am a Disco Dancer (Mashup)        0.0153         0.689   

   ...   key  liveness  loudness  mode  speechiness    tempo  time_signature  \
0  ...   8.0    0.1000    -5.994   0.0       0.1030  110.997             4.0   
1  ...  10.0    0.0951    -5.674   0.0       0.0952  164.995             3.0   
2  ...   2.0    0.0831    -8.937   0.0       0.1530  169.996             4.0   
3  ...   7.0    0.1240   -11.104   1.0       0.4450  169.996             4.0   
4  ...   7.0    0.3450    -9.637   1.0       0.1580  128.961             4.0   

   valence                                          track_url  language  
0    0.459  https://open.spotify.com/track/2r0ROhr7pRN4MXD...     Tamil  
1    0.821  https://open.spotify.com/track/4I38e6Dg52a2o2a...     Tamil  
2    0.598  https://open.spotify.com/track/59NoiRhnom3lTeR...     Tamil  
3    0.362  https://open.spotify.com/track/5uUqRQd385pvLxC...     Tamil  
4    0.593  https://open.spotify.com/track/1KaBRg2xgNeCljm...     Tamil  

[5 rows x 23 columns]

In [85]:
# Identify duplicates based on the 'spotify_id' column
duplicates = df_universal_songs[df_universal_songs.duplicated(subset=['spotify_id'], keep=False)]

# Display the duplicates
print(duplicates)


                     spotify_id                               name  \
0        2plbrEY59IikOBgBGLjaoe                   Die With A Smile   
1        3sK8wGT43QFpWrvNQsrQya                               DtMF   
2        4wJ5Qq0jBN4ajy7ouZIV1c                               APT.   
3        2lTm559tuIvatlT1u0JYG2                  BAILE INoLVIDABLE   
4        6dOtVTDdiauQNBQEDOtlAB                 BIRDS OF A FEATHER   
...                         ...                                ...   
1652689  0AYt6NMyyLd0rLuvr0UkMH          Slime You Out (feat. SZA)   
1652690  2Gk6fi0dqt91NKvlzGsmm7  SAY MY GRACE (feat. Travis Scott)   
1652691  26b3oVLrRUaaybJulow9kz                             People   
1652692  5ydjxBSUIDn26MFzU3asP4                         Rainy Days   
1652693  59NraMJsLaMCVtwXTSia8i                              Prada   

                             artists  daily_rank  daily_movement  \
0              Lady Gaga, Bruno Mars           1               1   
1                      

In [86]:
# Drop duplicate rows based on 'spotify_id'
df_universal_songs = df_universal_songs.drop_duplicates(subset=['spotify_id'])

In [87]:
len(df_universal_songs)

19764

In [88]:
df_spotify_songs = pd.read_csv("/Users/pratikhotchandani/Downloads/Github/DS_5500/spotify_top_songs_audio_features.csv")

In [89]:
len(df_spotify_songs)

6513

In [90]:
df_spotify_songs.head()

id                artist_names  \
0  000xQL6tZNLJzIrtIgxqSl         ZAYN, PARTYNEXTDOOR   
1  003eoIwxETJujVWmNFMoZy                Alessia Cara   
2  003vvx7Niy0yvhvHt4a68B                 The Killers   
3  00B7TZ0Xawar6NZ00JFomN  Cardi B, Chance the Rapper   
4  00Blm7zeNqgYLPtW6zg8cj     Post Malone, The Weeknd   

                             track_name              source    key   mode  \
0  Still Got Time (feat. PARTYNEXTDOOR)   RCA Records Label      G  Major   
1                         Growing Pains  Def Jam Recordings  C#/Db  Minor   
2                        Mr. Brightside      Island Records  C#/Db  Major   
3   Best Life (feat. Chance The Rapper)        Atlantic/KSR      A  Major   
4       One Right Now (with The Weeknd)    Republic Records  C#/Db  Major   

  time_signature  danceability  energy  speechiness  acousticness  \
0        4 beats         0.748   0.627       0.0639       0.13100   
1        4 beats         0.353   0.755       0.7330       0.08220   
2        4 beats         0.352   0.911       0.0747       0.00121   
3        4 beats         0.620   0.625       0.5530       0.28700   
4        4 beats         0.687   0.781       0.0530       0.03610   

   instrumentalness  liveness  valence  loudness    tempo  duration_ms  \
0               0.0    0.0852    0.524    -6.029  120.963       188491   
1               0.0    0.3900    0.437    -6.276  191.153       193680   
2               0.0    0.0995    0.236    -5.230  148.033       222973   
3               0.0    0.3140    0.665    -7.438  167.911       284856   
4               0.0    0.0755    0.688    -4.806   97.014       193507   

   weeks_on_chart    streams  
0              17  107527761  
1               2    9944865  
2             125  512388123  
3               2   11985346  
4              30  301860377

In [91]:
# Rename specific columns
df_spotify_songs.rename(columns={'id': 'spotify_id', 'track_name': 'name', 'artist_names': 'artists'}, inplace=True)

In [92]:
df_spotify_songs = df_spotify_songs.drop_duplicates(subset=['spotify_id'])
len(df_spotify_songs)

6513

In [93]:
df_tiktok = pd.read_csv("/Users/pratikhotchandani/Downloads/Github/DS_5500/tiktok.csv")

In [94]:
df_tiktok.head()

track_id                     track_name  \
0  6kVuF2PYLuvl9T85XjNbaO        Lay It Down Gmix - Main   
1  1RGIjMFMgJxkZHMDXVYzOJ         Bartender (feat. Akon)   
2  1RGIjMFMgJxkZHMDXVYzOJ         Bartender (feat. Akon)   
3  1dIWPXMX4kRHj6Dt2DStUQ   Chosen (feat. Ty Dolla $ign)   
4  4QVS8YCpK71R4FsxSMCjhP  Tie Me Down (with Elley Duhé)   

                artist_id artist_name                album_id  duration  \
0  1Xfmvd48oOhEWkscWyEbh9       Lloyd  43C6GVlhXG4KfZuEbxty3r    302186   
1  3aQeKQSyrW4qWr35idm0cy      T-Pain  6CrSEKCF4TYrbSIitegb3h    238800   
2  3aQeKQSyrW4qWr35idm0cy      T-Pain  6CrSEKCF4TYrbSIitegb3h    238800   
3  4qXC0i02bSFstECuXP2ZpL       Blxst  7AwrgenNcTAJlJF3pKL0Qr    161684   
4  2ZRQcIgzPCVaT9XKhXZIzh     Gryffin  69t8rpgBN1ov5kCU6LDMuR    218295   

  release_date  popularity  danceability  energy  ...  speechiness  \
0   2011-01-01          28         0.597   0.800  ...       0.3120   
1   2007-06-05          75         0.832   0.391  ...       0.0628   
2   2007-06-05          75         0.832   0.391  ...       0.0628   
3   2020-12-04          76         0.571   0.767  ...       0.2870   
4   2018-08-03          72         0.548   0.839  ...       0.0644   

   acousticness  instrumentalness  liveness  valence    tempo  \
0        0.0461               0.0    0.1800    0.565  155.932   
1        0.0564               0.0    0.2240    0.436  104.961   
2        0.0564               0.0    0.2240    0.436  104.961   
3        0.3360               0.0    0.0809    0.605   93.421   
4        0.1350               0.0    0.1020    0.314   98.932   

              playlist_id           playlist_name  duration_mins         genre  
0  6kVuF2PYLuvl9T85XjNbaO  6kVuF2PYLuvl9T85XjNbaO       5.036433  TIKTOK DANCE  
1  1RGIjMFMgJxkZHMDXVYzOJ  1RGIjMFMgJxkZHMDXVYzOJ       3.980000  TIKTOK DANCE  
2  1RGIjMFMgJxkZHMDXVYzOJ  1RGIjMFMgJxkZHMDXVYzOJ       3.980000  TIKTOK DANCE  
3  1dIWPXMX4kRHj6Dt2DStUQ  1dIWPXMX4kRHj6Dt2DStUQ       2.694733  TIKTOK DANCE  
4  4QVS8YCpK71R4FsxSMCjhP  4QVS8YCpK71R4FsxSMCjhP       3.638250  TIKTOK DANCE  

[5 rows x 23 columns]

In [95]:
# Rename specific columns
df_tiktok.rename(columns={'track_id': 'spotify_id', 'track_name': 'name', 'artist_name': 'artists'}, inplace=True)

In [96]:
df_tiktok = df_tiktok.drop_duplicates(subset=['spotify_id'])
len(df_tiktok)

3560

In [97]:
column_names_universal = list(df_universal_songs.columns)
column_names_universal

['spotify_id',
 'name',
 'artists',
 'daily_rank',
 'daily_movement',
 'weekly_movement',
 'country',
 'snapshot_date',
 'popularity',
 'is_explicit',
 'duration_ms',
 'album_name',
 'album_release_date',
 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo',
 'time_signature']

In [98]:
columns_to_drop = ['daily_rank', 'daily_movement', 'weekly_movement','country','snapshot_date','popularity','is_explicit','duration_ms','album_name','album_release_date']
df_universal_songs = df_universal_songs.drop(columns=columns_to_drop)
column_names_universal = list(df_universal_songs.columns)
column_names_universal

['spotify_id',
 'name',
 'artists',
 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo',
 'time_signature']

In [99]:
column_names_spotify = list(df_spotify_songs.columns)
column_names_spotify

['spotify_id',
 'artists',
 'name',
 'source',
 'key',
 'mode',
 'time_signature',
 'danceability',
 'energy',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'loudness',
 'tempo',
 'duration_ms',
 'weeks_on_chart',
 'streams']

In [101]:
columns_to_drop = ['source', 'duration_ms', 'weeks_on_chart','streams']
df_spotify_songs = df_spotify_songs.drop(columns=columns_to_drop)
column_names_spotify = list(df_spotify_songs.columns)
column_names_spotify

['spotify_id',
 'artists',
 'name',
 'key',
 'mode',
 'time_signature',
 'danceability',
 'energy',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'loudness',
 'tempo']

In [102]:
column_names_tiktok = list(df_tiktok.columns)
column_names_tiktok

['spotify_id',
 'name',
 'artist_id',
 'artists',
 'album_id',
 'duration',
 'release_date',
 'popularity',
 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo',
 'playlist_id',
 'playlist_name',
 'duration_mins',
 'genre']

In [103]:
columns_to_drop = ['artist_id', 'album_id', 'duration','release_date','playlist_id','playlist_name','duration_mins','genre']
df_tiktok = df_tiktok.drop(columns=columns_to_drop)
column_names_tiktok = list(df_tiktok.columns)
column_names_tiktok

['spotify_id',
 'name',
 'artists',
 'popularity',
 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo']

In [108]:
unique_universal_ids = df_universal_songs['spotify_id'].unique().tolist()
unique_universal_ids

['2plbrEY59IikOBgBGLjaoe',
 '3sK8wGT43QFpWrvNQsrQya',
 '4wJ5Qq0jBN4ajy7ouZIV1c',
 '2lTm559tuIvatlT1u0JYG2',
 '6dOtVTDdiauQNBQEDOtlAB',
 '5TFD2bmFKGhoCRbX61nXY5',
 '7ne4VBA60CxGM75vw0EYad',
 '59D4DOkspUbWyMmbAPQkxZ',
 '35ISBknsCeZQtq66xABI9g',
 '7tI8dRuH2Yc6RuoTjxo4dU',
 '2CGNAOSuO1MEFCbBRgUzjd',
 '2262bWmqomIaJXwCRHr13j',
 '7d6yK8v8J484SWH5prIQiE',
 '6J5kc12BW5HuP3d7C3vvx8',
 '1Es7AUAhQvapIcoh3qMKDL',
 '2HRqTpkrJO5ggZyyK6NPWz',
 '63EQbNhgaih4yg5412usWG',
 '3QaPy1KgI7nu9FJEQUgn6h',
 '5WEF0icHWmAZBBMglBd599',
 '0WbMK4wrZ1wFSty9F7FCgu',
 '6WatFBLVB0x077xWeoVc2k',
 '3be9ACTxtcL6Zm4vJRUiPG',
 '0l0vcZMU7AOeQmUIREoI2U',
 '7CyPwkp0oE8Ro9Dd5CUDjW',
 '6s8WSX1MxNThrot8ThI6fG',
 '0LTwdL5yZ6YOTEGUQPFuSN',
 '6jbYpRPTEFl1HFKHk1IC0m',
 '17phhZDn6oGtzMe56NuWvj',
 '4MUrUT38U98olfnyzMBCzm',
 '3U3hFkMr0Q90pD24EkE3Pr',
 '2zOmS55knKWSgScYPTNmGQ',
 '2QjOHCTQ1Jl3zawyYOpxh6',
 '5XeFesFbtLpXzIVDNQP22n',
 '5G2f63n7IPVPPjfNIGih7Q',
 '3hRV0jL3vUpRrcy398teAU',
 '0aB0v4027ukVziUGwVGYpG',
 '3AJwUDP919kvQ9QcozQPxg',
 

In [109]:
unique_spotify_ids = df_spotify_songs['spotify_id'].unique().tolist()
unique_spotify_ids

['000xQL6tZNLJzIrtIgxqSl',
 '003eoIwxETJujVWmNFMoZy',
 '003vvx7Niy0yvhvHt4a68B',
 '00B7TZ0Xawar6NZ00JFomN',
 '00Blm7zeNqgYLPtW6zg8cj',
 '00EPIEnX1JFjff8sC6bccd',
 '00ETaeHUQ6lops3oWU1Wrt',
 '00ZKeP47bZtswtANkvxz2j',
 '00gpGR84M27moP7AFuqHIx',
 '00imgaPlYRrMGn9o83hfmk',
 '00ko9WaS4jOX1kEk3gvHjf',
 '00lNx0OcTJrS3MKHcB80HY',
 '00rxTS5TKv3oXpDA6nWfBi',
 '00syWkRGIVQvYsg2OwfBUw',
 '017PF4Q3l4DBUiWoXk4OWT',
 '01DOm09B2AbiTARAbXVI6g',
 '01K4zKU104LyJ8gMb7227B',
 '01Qb6dN5suQNu87VJ1mIpf',
 '01TnVDiet1DFTsyWKUKovl',
 '01dmH2IPkrLNWIMPNXlreE',
 '01hEpx8jnOG4q7nDK14Rhi',
 '01k24g94i1JvkFLQmVEdCd',
 '01ppKlDRCmPpusO3yNrSRY',
 '01qFKNWq73UfEslI0GvumE',
 '01uk7IzZyFNfQTDBXxx6NB',
 '01vR5Z7AHinp9D2wu2BNQS',
 '01vv2AjxgP4uUyb8waYO5Y',
 '01z2fBGB8Hl3Jd3zXe4IXR',
 '020wG4EHTwBy5h2rbpxqvK',
 '021lWIHJrURZrmvYQ1xcM8',
 '0247StOpd3AkeBQzANX4Zf',
 '02FaKXXL7KUtRc7K0k54tL',
 '02LZ1fGJE8fB56yTlCVMEV',
 '02MWAaffLxlfxAUY7c5dvx',
 '02RCbjb9czvQKNGBmEmWob',
 '02VBYrHfVwfEWXk5DXyf0T',
 '02WacdrRpm4zlP8H7X6bnQ',
 

In [110]:
common_ids = set(unique_spotify_ids) & set(unique_universal_ids)

# If you want to print the common IDs
print(common_ids)

# To check if there are any common IDs (Boolean value)
has_common_ids = len(common_ids) > 0
print(has_common_ids)


{'1vvcEHQdaUTvWt0EIUYcFK', '5mHdCZtVyb4DcJw8799hZp', '6DUdDIRgLqCGq1DwkNWQTN', '5cF0dROlMOK5uNZtivgu50', '3B54sVLJ402zGa6Xm4YGNe', '1N8TTK1Uoy7UvQNUazfUt5', '0u8rZGtXJrLtiSe34FPjGG', '4MjDJD8cW7iVeWInc2Bdyj', '5QVVjX0ZItqlVpEuVCM9Yg', '214nt20w5wOxJnY462klLw', '1mH6tZC6iXIHPI6EixCwxw', '23RoR84KodL5HWvUTneQ1w', '59hBR0BCtJsfIbV9VzCVAp', '5tEaVciE2GnR28aN6W9cLS', '5QDLhrAOJJdNAmCTJ8xMyW', '51qYdMBxt9mFbtZBzq1Yfi', '4HYDUMY0xSpeBr0AMY9cUz', '4SqWKzw0CbA05TGszDgMlc', '4so0Wek9Ig1p6CRCHuINwW', '3TSLqZssCoCdDlMhCJ08XW', '5LrN7yUQAzvthd4QujgPFr', '5wANPM4fQCJwkGd4rN57mH', '59NraMJsLaMCVtwXTSia8i', '75JFxkI2RXiU7L9VXzMkle', '27RYrbL6S02LNVhDWVl38b', '0RiRZpuVRbi7oqRdSMwhQY', '741UUVE2kuITl0c6zuqqbO', '5Se32hEA9raeboZerywxka', '7uoFMmxln0GPXQ0AcCBXRq', '4r9jkMEnArtWGH2rL2FZl0', '1JgknGBbrfmEHeOZH051SS', '6IPwKM3fUUzlElbvKw2sKl', '0BFQC00Y58JtjMyTXOJazO', '0CYTGMBYkwUxrj1MWDLrC5', '0cGpKQ8RxxLOFqhktEURIj', '0kdqcbwei4MDWFEX5f33yG', '5LhI8m7n8xlQmNe3b7Rwis', '0Jiaz0O4AqnJICa9PxHhaR', '4ZnkygoWLz

In [111]:
df_combined = pd.DataFrame()

In [112]:
import pandas as pd

# Concatenate the dataframes
df_combined = pd.concat([df_universal_songs, df_spotify_songs], ignore_index=True)

# Drop duplicates, keeping the first occurrence (which will be from df1)
result_df = df_combined.drop_duplicates(subset='spotify_id', keep='first')


In [113]:
len(result_df)

25411

In [114]:
result_df

spotify_id                                         name  \
0      2plbrEY59IikOBgBGLjaoe                             Die With A Smile   
1      3sK8wGT43QFpWrvNQsrQya                                         DtMF   
2      4wJ5Qq0jBN4ajy7ouZIV1c                                         APT.   
3      2lTm559tuIvatlT1u0JYG2                            BAILE INoLVIDABLE   
4      6dOtVTDdiauQNBQEDOtlAB                           BIRDS OF A FEATHER   
...                       ...                                          ...   
26272  7zgqtptZvhf8GEmdsM2vp2                             ...Ready For It?   
26273  7zjEyeBsaw9gV0jofJLfOM  Livin It Up (with Post Malone & A$AP Rocky)   
26274  7zl7kehxesNEo2pYkKXTSe  Killer (feat. Jack Harlow & Cordae) - Remix   
26275  7zvfDihYiJ8RQ1nRcpKBF5                                   Mr. Morale   
26276  7zxRMhXxJMQCeDDg0rKAVo                                     Some Way   

                                   artists  danceability   energy key  \
0                    Lady Gaga, Bruno Mars         0.521  0.59200   6   
1                                Bad Bunny         0.625  0.13100   7   
2                         ROSÉ, Bruno Mars         0.777  0.78300   0   
3                                Bad Bunny         0.168  0.00332  10   
4                            Billie Eilish         0.747  0.50700   2   
...                                    ...           ...      ...  ..   
26272                         Taylor Swift         0.615  0.77900   D   
26273  Young Thug, A$AP Rocky, Post Malone         0.767  0.31300   G   
26274          Eminem, Jack Harlow, Cordae         0.924  0.73000   B   
26275          Kendrick Lamar, Tanna Leone         0.727  0.53000   A   
26276                      NAV, The Weeknd         0.744  0.71500   C   

       loudness   mode  speechiness  acousticness  instrumentalness  liveness  \
0        -7.777      0       0.0304        0.3080          0.000000    0.1220   
1       -27.405      0       0.0717        0.1770          0.218000    0.0807   
2        -4.477      0       0.2600        0.0283          0.000000    0.3550   
3       -46.113      1       0.0615        0.1920          0.790000    0.1120   
4       -10.171      1       0.0358        0.2000          0.060800    0.1170   
...         ...    ...          ...           ...               ...       ...   
26272    -6.454  Major       0.1350        0.0665          0.000000    0.1550   
26273   -12.059  Major       0.0798        0.8380          0.000000    0.1050   
26274    -6.125  Minor       0.2740        0.0525          0.000000    0.1130   
26275    -8.616  Major       0.3120        0.2830          0.000028    0.3920   
26276    -6.103  Major       0.3510        0.1010          0.000000    0.0919   

       valence    tempo time_signature  
0        0.535  157.969              3  
1        0.032  136.020              4  
2        0.939  149.027              4  
3        0.219  119.387              3  
4        0.438  104.978              4  
...        ...      ...            ...  
26272    0.453  160.000        4 beats  
26273    0.765   82.582        4 beats  
26274    0.848   99.010        4 beats  
26275    0.258  174.025        3 beats  
26276    0.340  154.962        4 beats  

[25411 rows x 15 columns]

In [115]:
import pandas as pd

# Concatenate the dataframes
df_combined = pd.concat([result_df, df_tiktok], ignore_index=True)

# Drop duplicates, keeping the first occurrence (which will be from df1)
result_df = df_combined.drop_duplicates(subset='spotify_id', keep='first')


In [116]:
len(result_df)

28451

In [117]:
result_df

spotify_id                               name  \
0      2plbrEY59IikOBgBGLjaoe                   Die With A Smile   
1      3sK8wGT43QFpWrvNQsrQya                               DtMF   
2      4wJ5Qq0jBN4ajy7ouZIV1c                               APT.   
3      2lTm559tuIvatlT1u0JYG2                  BAILE INoLVIDABLE   
4      6dOtVTDdiauQNBQEDOtlAB                 BIRDS OF A FEATHER   
...                       ...                                ...   
28964  1CkTEh21SLHbiAj2qy8wc1                             Ma Boy   
28966  03f7xZmt2wHCIDJBFPK8G4                               Soap   
28967  4kIpBfvK44bxqX7zo8K1oP                                 시작   
28968  1lNHWPDvKEbamKezpLq7HW                no song without you   
28969  7gd01LMH2gBcoDngSt8sq9  Next To You (feat. Justin Bieber)   

                     artists  danceability   energy key  loudness mode  \
0      Lady Gaga, Bruno Mars         0.521  0.59200   6    -7.777    0   
1                  Bad Bunny         0.625  0.13100   7   -27.405    0   
2           ROSÉ, Bruno Mars         0.777  0.78300   0    -4.477    0   
3                  Bad Bunny         0.168  0.00332  10   -46.113    1   
4              Billie Eilish         0.747  0.50700   2   -10.171    1   
...                      ...           ...      ...  ..       ...  ...   
28964               SISTAR19         0.751  0.76700   7    -4.752    0   
28966       Melanie Martinez         0.595  0.68900   6    -6.107    0   
28967                   Gaho         0.591  0.81800   4    -3.532    1   
28968                  HONNE         0.788  0.47300   0   -12.744    1   
28969            Chris Brown         0.670  0.67600   8    -4.725    1   

       speechiness  acousticness  instrumentalness  liveness  valence  \
0           0.0304        0.3080          0.000000    0.1220    0.535   
1           0.0717        0.1770          0.218000    0.0807    0.032   
2           0.2600        0.0283          0.000000    0.3550    0.939   
3           0.0615        0.1920          0.790000    0.1120    0.219   
4           0.0358        0.2000          0.060800    0.1170    0.438   
...            ...           ...               ...       ...      ...   
28964       0.0577        0.3530          0.000003    0.2890    0.768   
28966       0.1480        0.2540          0.000005    0.1290    0.213   
28967       0.0730        0.1720          0.000000    0.1260    0.574   
28968       0.0328        0.4890          0.254000    0.1090    0.810   
28969       0.0370        0.0201          0.000000    0.1100    0.349   

         tempo time_signature  popularity  
0      157.969              3         NaN  
1      136.020              4         NaN  
2      149.027              4         NaN  
3      119.387              3         NaN  
4      104.978              4         NaN  
...        ...            ...         ...  
28964   89.930            NaN        53.0  
28966  168.112            NaN        68.0  
28967  108.107            NaN        67.0  
28968  105.429            NaN        64.0  
28969  114.974            NaN        70.0  

[28451 rows x 16 columns]

In [121]:
result_df.to_csv("final_merged_data.csv")

#### Statistics Summary

In [118]:
# Get statistical summary
summary = result_df.describe()

print(summary)

       danceability        energy      loudness   speechiness  acousticness  \
count  28451.000000  28451.000000  28451.000000  28451.000000  28451.000000   
mean       0.682499      0.645936     -6.940185      0.124638      0.260524   
std        0.141391      0.171364      3.094029      0.115597      0.248711   
min        0.000000      0.001740    -54.341000      0.000000      0.000003   
25%        0.594000      0.538000     -8.240000      0.043400      0.055000   
50%        0.701000      0.662000     -6.461000      0.072400      0.177000   
75%        0.785000      0.772000     -5.006500      0.172000      0.407000   
max        0.990000      0.999000      2.619000      0.966000      0.996000   

       instrumentalness      liveness       valence         tempo   popularity  
count      28451.000000  28451.000000  28451.000000  28451.000000  3040.000000  
mean           0.027170      0.182225      0.526853    122.165635    46.027303  
std            0.127195      0.138835      0.

#### Missing values

In [119]:
# Check for missing values
missing_values = result_df.isnull().sum()

# Display columns with missing values
print("Missing Values per Column:")
print(missing_values[missing_values > 0])


Missing Values per Column:
name                 10
artists               9
time_signature     3040
popularity        25411
dtype: int64


#### Outliers

In [120]:
from scipy.stats import zscore

# Compute Z-scores
z_scores = result_df.select_dtypes(include=['float64', 'int64']).apply(zscore)

# Identify outliers
outliers = (z_scores > 3) | (z_scores < -3)
print("Outliers per column:")
print(outliers.sum())


Outliers per column:
danceability        140
energy              180
loudness            378
speechiness         379
acousticness          0
instrumentalness    784
liveness            714
valence               0
tempo                35
popularity            0
dtype: int64


Okay now we have the data merged from the three different datasets from kaggle. We can now answer questions from the assignment.

## 1. Source:

Where did the data come from (source)?
--> Kaggle

What was the purpose of collecting this data?
--> This data will be used/stored as RAG data that a chatbot could use to suggest songs based on the features/mood identified by the Vision model. Instead of chatbot, we can use sql queries or pandas queries to suggest the song based on the mood identified.

Is the data publicly available, or are there restrictions on its use? please provide links 
-> Yes, the data is publicly available, although spotify has deprecated its api (getting tracks audio features), we can use other open source models to generate these features.

## 2. Structure and Metadata:

What are the key features (columns) in your dataset, and what do they represent?
--> 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo',
 'time_signature'

What is the size of your dataset? Provide statistical summaries and key metrics about your dataset.
--> total 28451 unique tracks.

Is there an API available for your dataset?
--> No, spotify recently deprecated its api to get the tracks features. (https://developer.spotify.com/blog/2024-11-27-changes-to-the-web-api)

## 3. Missing Data Analysis

Are there any missing values? If so, how are they represented?

Yes, the data contains missing values in the following columns:

name: 10 missing values.
artists: 9 missing values.
time_signature: 3040 missing values.
popularity: 25,411 missing values.

What strategies can be used to handle missing data?
Recommendation Based on Your Data:
For name and artists: These are critical fields; you can check if the missing rows are essential. If so, try to fill the gaps using additional data or drop rows if they are few.
For time_signature: This may be replaced with the most frequent value (e.g., 4).
For popularity: If this feature is important, you can impute missing values using the mean or median. Otherwise, consider removing the column.


## 4. Anomalies

Are there outliers or anomalies? do they indicate errors?
Outliers may or may not indicate errors, depending on the context of the data. Let’s break it down:

Danceability, Energy, and Tempo:
These features are numeric and typically range between 0 and 1 or specific bounds for music datasets.
Outliers could represent unusual or extreme songs (e.g., highly energetic or slow-tempo tracks) rather than errors.
Loudness:
Loudness is typically measured in decibels (dB) and can have negative values.
Outliers may reflect extremely quiet or unusually loud tracks.
Errors could occur if unrealistic loudness values (e.g., >0 dB or <-60 dB) are present.
Speechiness and Liveness:
These values indicate specific audio characteristics.
Outliers could represent tracks with an unusual amount of speech or live audience sounds.
Verify if the values are within the valid range (e.g., 0–1).
Instrumentalness:
Outliers may represent rare cases like entirely instrumental or highly vocal tracks.
Unusually high counts of outliers could indicate the need for further inspection.
Acousticness and Valence:
No outliers were detected. These features appear consistent.
Popularity:
No outliers were found, but ensure that values fall within the valid range (e.g., 0–100).

How consistent is the data across features and observations?
Data Type Consistency: Check if all features have consistent data types (e.g., numeric for danceability, energy, etc.). Inconsistent types can indicate errors.
print(df.dtypes)
Range Consistency:
Features like danceability, energy, speechiness, and liveness are expected to be in the range [0, 1].
Loudness should typically range between -60 dB and 0 dB.
Tempo values are usually positive.
print(df.describe())
Duplications: Duplicate rows or entries with the same spotify_id might indicate inconsistencies.
